builtin/mainmenu/tab_online.lua

```diff
@@ -264,19 +265,94 @@ end
 
 --------------------------------------------------------------------------------
 
+local function parse_search_input(input)
+	if not input:find("%S") then
+		return -- Return nil if nothing to search for
+	end
+
+	-- Search is not case sensitive
+	input = input:lower()
+
+	local query = {keywords = {}, mods = {}, players = {}}
+
+	-- Process quotation enclosed parts
+	input = input:gsub('(%S?)"([^"]*)"(%S?)', function(before, match, after)
+		if before == "" and after == "" then -- Also have be separated by spaces
+			table.insert(query.keywords, match)
+			return " "
+		end
+		return before..'"'..match..'"'..after
+	end)
+
+	-- Separate by space characters and handle special prefixes
+	-- (words with special prefixes need an exact match and none of them can contain spaces)
+	for word in input:gmatch("%S+") do
+		local mod = word:match("^mod:(.*)")
+		table.insert(query.mods, mod)
+		local player = word:match("^player:(.*)")
+		table.insert(query.players, player)
+		local game = word:match("^game:(.*)")
+		query.game = query.game or game
+		if not (mod or player or game) then
+			table.insert(query.keywords, word)
+		end
+	end
+
+	return query
+end
+
+-- Returns false if filter doesn't match (special prefixes like "mod:", "game:", "player:")
+-- otherwise returns true, name_matches, description_matches
+local function matches_query(server, query)
+	-- Check if mods found
+	local mods_lower = {}
+	for j, mod in ipairs(server.mods or {}) do
+		mods_lower[j] = mod:lower()
+	end
+	for _, mod in ipairs(query.mods) do
+		if table.indexof(mods_lower, mod) < 0 then
+			return false
+		end
+	end
+
+	-- Check if players found
+	local clients_list_lower = {}
+	for j, player in ipairs(server.clients_list or {}) do
+		clients_list_lower[j] = player:lower()
+	end
+	for _, player in ipairs(query.players) do
+		if table.indexof(clients_list_lower, player) < 0 then
+			return false
+		end
+	end
+
+	-- Check if game matches
+	if query.game and query.game ~= server.gameid then
+		return false
+	end
+
+	local name_matches, description_matches = true, true
+
+	-- Check if keyword found
+	for _, keyword in ipairs(query.keywords) do
+		name_matches = name_matches and not not
+				(server.name or ""):lower():find(keyword, 1, true)
+		description_matches = description_matches and not not
+				(server.description or ""):lower():find(keyword, 1, true)
+	end
+
+	return true, name_matches, description_matches
```

Comment:
```suggestion
	return name_matches or description_matches
```

makes more sense IMO, then the below logic simplifies a bit

Reply:
This does not work since the server sorting priority depends on `name_matches`
https://github.com/minetest/minetest/blob/7450ebd5e484b1b196a68a0013463d9777927e75/builtin/mainmenu/tab_online.lua#L366-L370

So, instead, I made the `matches_query` function return the priority change, hopefully this is better.